### PR TITLE
Get "## commits to master since this release" working

### PR DESCRIPTION
**Pull request by umireon**
GitHub Releases has an awesome feature that shows how many commits are created since the release.
Example: [github/hub releases](https://github.com/github/hub/releases)

hub 2016-11-30 01-36-23

However, the project which uses semantic-release seems to not to have this awesomeness.
Example: [semantic-release/semantic-release releases](https://github.com/semantic-release/semantic-release/releases)

semantic-release 2016-11-30 01-37-55

## Description

`target_commitish` in GitHub Releases has two different meanings:
* Target commit that a new associated tag will be created on
* Target branch to compute how many "commits to since this release"

`target_commitish` is usually the default branch (aka `master`), because the
distance between the release and the latest branch is the primary concern.

Before this change, `target_commitish` was left to be a hash and the
feature of GitHub Releases that shows how much time passed since the
release is ruined, because no tracking branch is given.

By this change, `target_commitish` is changed to be the default branch
given in the configuration (`options.branch`) via edit a release API after
the release and tag are created via crate a release API.

### CODE REVIEW

**Comment by umireon** [comment]
According to [API doc](https://developer.github.com/v3/repos/releases/#input), `target_commitish` is described to be usually `master` (or the default branch) and that is almost always true.
This can be confirmed by this command: `curl -fsSL https://api.github.com/repos/github/hub/releases | grep '"target_commitish"'`

```
$ curl -u umireon:$GH_TOKEN -fsSL https://api.github.com/repos/github/hub/releases | grep '"target_commitish"'
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "2.2-stable",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "2.2-stable",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
    "target_commitish": "master",
```

But `semantic-release`ed repos has weired `target_commitish`es...

```
$ curl -fsSL https://api.github.com/repos/semantic-release/semantic-release/releases | grep '"target_commitish"'
    "target_commitish": "73138f8a8f15b207d6553864c17917433be38800",
    "target_commitish": "f239f2f402551bb210c3ae0475e0349906829b46",
    "target_commitish": "2be6e5d34a4c89a74a8a0bd40a6706d3705be448",
    "target_commitish": "a5cb9ea7f7a66657f95f0d913126e812adb6d9f8",
    "target_commitish": "0a89265ae76b5672465f04caae0676104bd28894",
    "target_commitish": "d9b95940dbfbbb321810cbcb35886d5eff330acd",
    "target_commitish": "4b64afefe4321bdf4c70cdfe81e00cbee2d2efd8",
    "target_commitish": "adc65e4950f6218345e73be9c959a209ead8d871",
    "target_commitish": "64926934894fb8ae69892638df522d5e31deed0d",
    "target_commitish": "e19b1952d174e3c7b41a87025686f2508551a126",
    "target_commitish": "676d2d6210f2bfc17a376c7eeb59bbf5dd02dd98",
    "target_commitish": "df5f8337e89de14bb1025847245ffe4a07e6b152",
    "target_commitish": "527c45fbf5c7a47ef30cd689208843b892f54033",
    "target_commitish": "a7270bb84c13e803ca851e9426c2cef5071a583b",
    "target_commitish": "ef1623210ba2b5b97edb4d76c497513ef729b3e9",
    "target_commitish": "5711e95c790a269a97d705b8b89d6292d0e093c0",
    "target_commitish": "03534e51bcc57c6a3379e885d1541efa777c965d",
    "target_commitish": "451378117df6d7015f5db1dacc99d9ade4ebc123",
    "target_commitish": "85b35aaed363a616d9144cae2c0e88ed9590511d",
    "target_commitish": "23221cc044c1498ee63f5b05d17b7ff527358b03",
    "target_commitish": "a16c36ed4d9fb1a10b84069f9239e63bb4605d7a",
    "target_commitish": "a2d6db2ce5511f99ffa7efed1cf3f8427464693a",
    "target_commitish": "b47311d5653065a67a59ea3869d5e93ed5ccc5b7",
    "target_commitish": "7ceac766892b6f4c311582efcdd3a2c86ae3dad2",
    "target_commitish": "55364ded401dbabbb867fde0cf2be01b4d2f7790",
    "target_commitish": "fb1aa80f263881c29ea84d5896b8a66ee95b390a",
    "target_commitish": "05490fe2290144fbc60ae3171184e4e05719898a",
    "target_commitish": "caribou",
    "target_commitish": "2690d30aa07958ed7e109cb6536046b3af391f68",
    "target_commitish": "1db531b51180a3d032024bbb5a0cf4e1bed1147e",
```

**Comment by umireon** [comment]
Broken `target_commitish`es can be fixed by this script.
It's working since [my repo](https://github.com/umireon/editorconfig-jxa/releases) was fixed by this.

```
ENDPOINT=https://api.github.com/repos/username/reponame/releases
GH_AUTH="username:password_or_token"
for id in $(curl -u "$GH_AUTH" -fsSL "$ENDPOINT" | jq '.[].id')
do
  curl -v \
    -u "$GH_AUTH" \
    -X  PATCH \
    -d '{"target_commitish": "master"}' \
    "$ENDPOINT/$id"
done
```

**Comment by boennemann** [comment]
Hey @umireon,

thanks for digging into this detail, I really like it 👍 
I'm happy to merge this, but I don't think that first creating and then editing the release is the most elegant solution.

What do you think about first using the git data API to [create a tag at the specified hash](https://developer.github.com/v3/git/tags/#create-a-tag-object) and then use the release API to create a release from that tag, including the release notes?


Best,
Stephan

**Comment by umireon** [comment]
Thank you, @boennemann !

I have changed the PR to use [Git Refs API](https://developer.github.com/v3/git/refs/#create-a-reference) to make a lightweight tag (the same as the current implementation does).

Note that the suggested API is not needed to call for creating a lighweight tag.

**Comment by gr2m** [comment]
Hello there @umireon! I’m sorry nobody got back to for so long 😞 I recently [joined as maintainer](https://github.com/semantic-release/semantic-release/issues/411) and would love to help get this merged and released. 

It still need some time to familiarise myself with the code base and the respective GitHub APIs used in this PR. I’ve seen you already added tests so that seems good. I’m not yet sure how I can test it myself before release, if you have any ideas, let me know :)

And in case we release it and run into any trouble, I’d appreciate to know that you are around to diagnose and fix eventual problems.

Thanks for your help 💐

**Comment by umireon** [comment]
Thank you for picking this up, @gr2m 

I have rebased the branch and confirmed my tests passed on Travis CI.

In addition, I have created a sandbox repository for testing.
https://github.com/umireon/semantic-release-sandbox-target-committish

Best regards,
Kaito

File: test/specs/post.js [hunk 1]
```diff
@@ -23,14 +23,14 @@ var defaultRelease = {
...
-  target_commitish: 'bar',
+  target_commitish: 'master',
   body: 'the log'
 }
 
 test('full post run', function (t) {
   t.test('in debug mode w/o token', function (tt) {
     post({
-      options: {debug: true},
+      options: {debug: true, branch: 'master'},
```
Comment (gr2m) [inline test/specs/post.js]: can you explain why this change was needed? Just want to make sure that merging this change won’t break something else, also want to better understand you changes :)

Comment (umireon) [inline test/specs/post.js]: This PR added the usage of `options.branch` in src/post.js, and its default value is `master`.
https://github.com/semantic-release/semantic-release#options

This change just provides the default value of `options.branch` for the test code.

**Comment by gr2m** [comment]
All right, thanks you very much, I’m going to rebase your PR on latest master and merge it once CI is happy 👍

**Comment by gr2m** [comment]
released as https://github.com/semantic-release/semantic-release/releases/tag/v7.0.1 (not yet on `@latest` on npm, install with `@next`)

**Comment by gr2m** [comment]
@umireon Hey is it possible that this change creates branche names with the same names as the git tags that get created? See https://github.com/semantic-release/semantic-release/issues/415 and https://github.com/semantic-release/semantic-release/tree/v7.0.1

Is there a way to avoid that?

**Comment by umireon** [comment]
I'm sorry but it's my mistake.

https://github.com/semantic-release/semantic-release/pull/335/files#diff-012dd2b56a863a41776f3a519e6aaa56R30

```
ref: 'refs/heads/v' + pkg.version,
```

should be

```
ref: 'refs/tags/v' + pkg.version,
```

**Comment by gr2m** [comment]
no worries at all, I haven’t found it when I reviewed, so it’s equally my fault. If you could send another PR that would be great :) No rush though

**Comment by umireon** [comment]
Will be fixed by #416
